skill-tree/src/test/java/com/RDS/skilltree/unitTests/EndorsementsApiUnitTests.java

```diff
@@ -0,0 +1,85 @@
+package com.RDS.skilltree.unitTests;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+
+import com.RDS.skilltree.dtos.RdsGetUserDetailsResDto;
+import com.RDS.skilltree.models.Endorsement;
+import com.RDS.skilltree.models.Skill;
+import com.RDS.skilltree.repositories.EndorsementRepository;
+import com.RDS.skilltree.services.EndorsementServiceImplementation;
+import com.RDS.skilltree.services.external.RdsService;
+import com.RDS.skilltree.viewmodels.EndorsementViewModel;
+import com.RDS.skilltree.viewmodels.RdsUserViewModel;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+public class EndorsementsApiUnitTests {
+
+    @Mock private EndorsementRepository endorsementRepository;
+
+    @Mock private RdsService rdsService;
+
+    @InjectMocks private EndorsementServiceImplementation endorsementService;
+
+    @BeforeEach
+    public void setup() {
+        MockitoAnnotations.openMocks(this); // Initialize mocks
+    }
+
+    @Test
+    @DisplayName("To test getAllEndorsementsBySkillId() method")
+    public void testGetAllEndorsementsBySkillId() {
+        List<Endorsement> endorsements = new ArrayList<>();
+
+        Integer skillId = 1;
+
+        // initializing a skill
+        Skill skill1 = new Skill();
+        skill1.setId(skillId);
+        skill1.setName("python");
+
+        // initializing an endorsement
+        Endorsement endorsement1 = new Endorsement();
+        endorsement1.setId(1);
+        endorsement1.setSkill(skill1);
+        endorsement1.setEndorserId("456");
+        endorsement1.setEndorseId("123");
+        endorsement1.setMessage("Please approve my endorsement for this skill");
+
+        endorsements.add(endorsement1);
+
+        RdsGetUserDetailsResDto rdsGetUserDetailsResDto1 = new RdsGetUserDetailsResDto();
+        RdsUserViewModel rdsUserViewModel1 = new RdsUserViewModel();
+        rdsUserViewModel1.setId("123");
+        rdsGetUserDetailsResDto1.setUser(rdsUserViewModel1);
+
+        RdsGetUserDetailsResDto rdsGetUserDetailsResDto2 = new RdsGetUserDetailsResDto();
+        RdsUserViewModel rdsUserViewModel2 = new RdsUserViewModel();
+        rdsUserViewModel2.setId("456");
+        rdsGetUserDetailsResDto2.setUser(rdsUserViewModel2);
+
+        // Mock behaviour
+        when(endorsementRepository.findBySkillId(skillId)).thenReturn(endorsements);
+        when(rdsService.getUserDetails(endorsement1.getEndorseId()))
+                .thenReturn(rdsGetUserDetailsResDto1);
+        when(rdsService.getUserDetails(endorsement1.getEndorserId()))
+                .thenReturn(rdsGetUserDetailsResDto2);
+
```

Comment:
Are we mocking the Db calls?

Reply:
Yes Prakash, correct

Reply:
Why are we mocking the db calls?

Reply:
As repository is a dependency for that method and we wanted to test in isolated space for unit tests, I have mocked it and provided pre defined data

Reply:
Hi @prakashchoudhary07 could you please let me know, for the above explanation regarding the reporsitory mocking, if any chnages need please let me know

Reply:
I won't not prefer mocking DB's, We should come up with a proper solution for this one

Reply:
In this case, if I don't have DB code, if queries are wrong also then tests will pass

Reply:
Hi @prakashchoudhary07 so this part of db validation is being covered as part of integration testing, so should we include in unit tests as well. please let me know?

cc @yesyash

Reply:
Well, I will still insist we do it. 
Now I am not sure what the test is being written for here. Is it a unit or integration?
So if you were testing the class why are you invoicing the db?
And if some behavior changes, then also since things are mocked, will tests pass without any issues right?

Reply:
And if you are only writing unit tests for a module, why  involve different modules here

Reply:
@prakashchoudhary07 

1. We are writing unit tests for the different methods in this pull request.
2. I have written unit tests for 4 methods in this pull request with different methods for each test.



> And if you are only writing unit tests for a module, why involve different modules here

  For the above question, for example, when writing unit tests for the **skills API getAll() method**, we are retrieving all the skills from the skill repository. So, the dependency for this method is the **skillRepository** class. To test the getAll method in isolation, I have mocked this class implementation. 

Based on the above conversation, do you think? should we include DB code as well? if yes should we use an in-memory database, as if we include the real db, as the data will be dynamic we should think of any strategy for this? could you please advise.

cc @yesyash